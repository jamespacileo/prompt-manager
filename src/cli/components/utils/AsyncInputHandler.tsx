import { Box, Text } from "ink";

import TextInput from "ink-text-input";
import { useState } from "react";

interface AsyncInputHandlerProps<T> {
  onSubmit: (value: string) => Promise<T>;
  onSuccess: (result: T) => void;
  placeholder?: string;
  errorMessage?: string;
}

export const AsyncInputHandler = <T,>({
  onSubmit,
  onSuccess,
  placeholder,
  errorMessage,
}: AsyncInputHandlerProps<T>) => {
  const [value, setValue] = useState("");
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  const handleSubmit = async () => {
    setIsLoading(true);
    setError(null);
    try {
      const result = await onSubmit(value);
      onSuccess(result);
    } catch (err) {
      setError(errorMessage || "An error occurred");
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <Box flexDirection="column">
      <TextInput
        value={value}
        onChange={setValue}
        onSubmit={handleSubmit}
        placeholder={placeholder}
      />
      {isLoading && <Text>Loading...</Text>}
      {error && <Text color="red">{error}</Text>}
    </Box>
  );
};
