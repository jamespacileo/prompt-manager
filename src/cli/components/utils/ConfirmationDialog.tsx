import { Box, Text, useInput } from "ink";

interface ConfirmationDialogProps {
	message: string;
	onConfirm: () => void;
	onCancel: () => void;
}

export const ConfirmationDialog: React.FC<ConfirmationDialogProps> = ({
	message,
	onConfirm,
	onCancel,
}) => {
	useInput((input) => {
		if (input === "y") {
			onConfirm();
		} else if (input === "n") {
			onCancel();
		}
	});

	return (
		<Box flexDirection="column">
			<Text color="yellow">{message}</Text>
			<Text>Press (Y) to confirm or (N) to cancel</Text>
		</Box>
	);
};
