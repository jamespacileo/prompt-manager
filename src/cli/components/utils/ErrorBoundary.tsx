import { Text } from "ink";
import React from "react";

class ErrorBoundary extends React.Component<
	{ children: React.ReactNode },
	{ hasError: boolean }
> {
	constructor(props: { children: React.ReactNode }) {
		super(props);
		this.state = { hasError: false };
	}

	static getDerivedStateFromError(_: Error) {
		return { hasError: true };
	}

	componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
		console.error("Uncaught error:", error, errorInfo);
	}

	render() {
		if (this.state.hasError) {
			return <Text color="red">Something went wrong. Please try again.</Text>;
		}

		return this.props.children;
	}
}

export default ErrorBoundary;
