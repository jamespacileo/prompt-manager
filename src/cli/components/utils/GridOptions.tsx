import { Box, Text } from "ink";
import type React from "react";

export interface Option {
	value: string;
	label: string;
	description?: string;
}

interface GridOptionsProps {
	options: Option[];
	selectedIndex?: number;
	maxDisplayedOptions?: number;
	descriptionMaxLength?: number;
	showNumbers?: boolean;
	highlightColor?: string;
}

const GridOptions: React.FC<GridOptionsProps> = ({
	options,
	selectedIndex = -1,
	maxDisplayedOptions = 10,
	descriptionMaxLength = 50,
	showNumbers = true,
	highlightColor = "green",
}) => {
	const truncate = (str: string, maxLength: number) =>
		str.length > maxLength ? `${str.slice(0, maxLength)}...` : str;

	return (
		<Box flexDirection="column">
			{options.slice(0, maxDisplayedOptions).map((option, index) => (
				<Box key={option.value}>
					<Text color={index === selectedIndex ? highlightColor : undefined}>
						{showNumbers && `${index + 1}. `}
						{option.label}
					</Text>
					{option.description && (
						<Text color="gray">
							{" "}
							- {truncate(option.description, descriptionMaxLength)}
						</Text>
					)}
				</Box>
			))}
		</Box>
	);
};

export default GridOptions;
