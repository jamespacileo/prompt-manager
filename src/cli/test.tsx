import { Box, Text, render, useFocus } from "ink";

import React from "react";

//you need this
process.stdin.resume();

function Focus() {
  return (
    <Box flexDirection="column" padding={1}>
      <Box marginBottom={1}>
        <Text>
          Press Tab to focus next element, Shift+Tab to focus previous element,
          Esc to reset focus.
        </Text>
      </Box>
      <Item label="First" />
      <Item label="Second" />
      <Item label="Third" />
    </Box>
  );
}

function Item({ label }: any) {
  const { isFocused } = useFocus();
  return (
    <Text>
      {label} {isFocused && <Text color="green">(focused)</Text>}
    </Text>
  );
}

render(<Focus />);
