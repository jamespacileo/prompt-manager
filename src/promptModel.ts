import { IPromptModel, IPromptModelRequired, IPromptModelStatic, IPromptInput, IPromptOutput, IAsyncIterableStream } from './types/interfaces';
import { JSONSchema7 } from 'json-schema';
import { generateText, generateObject, streamText } from 'ai';
import { openai } from '@ai-sdk/openai';
import { z } from 'zod';
import { PromptFileSystem } from './promptFileSystem';

export class PromptModel implements IPromptModel {
  name: string;
  category: string;
  description: string;
  version: string;
  template: string;
  parameters: string[];
  defaultModelName?: string;
  metadata: {
    created: string;
    lastModified: string;
  };
  configuration: {
    modelName: string;
    temperature: number;
    maxTokens: number;
    topP: number;
    frequencyPenalty: number;
    presencePenalty: number;
    stopSequences: string[];
  };
  outputType: 'structured' | 'plain';
  inputSchema: JSONSchema7;
  outputSchema: JSONSchema7;
  fileSystem: PromptFileSystem;
  _isSaved: boolean = false;
  isLoadedFromStorage: boolean = false;

  constructor(promptData: IPromptModelRequired, fileSystem?: PromptFileSystem) {
    this.name = promptData.name;
    this.category = promptData.category;
    this.description = promptData.description;
    this.template = promptData.template;
    this.parameters = promptData.parameters;
    this.inputSchema = promptData.inputSchema;
    this.outputSchema = promptData.outputSchema;
    this.fileSystem = fileSystem ?? new PromptFileSystem();
    this.version = '1.0.0';
    this.metadata = { created: new Date().toISOString(), lastModified: new Date().toISOString() };
    this.outputType = 'plain';
    this.configuration = this.initializeConfiguration();
  }

  private initializeConfiguration(): {
    modelName: string;
    temperature: number;
    maxTokens: number;
    topP: number;
    frequencyPenalty: number;
    presencePenalty: number;
    stopSequences: string[];
  } {
    return {
      modelName: this.defaultModelName || 'default-model',
      temperature: 0.7,
      maxTokens: 100,
      topP: 1,
      frequencyPenalty: 0,
      presencePenalty: 0,
      stopSequences: [],
    };
  }


  validateInput(input: IPromptInput): boolean {
    // Implement input validation logic using this.inputSchema
    return true; // Placeholder
  }

  validateOutput(output: IPromptOutput): boolean {
    // Implement output validation logic using this.outputSchema
    return true; // Placeholder
  }

  format(inputs: IPromptInput): string {
    let formattedContent = this.template;
    for (const [key, value] of Object.entries(inputs)) {
      formattedContent = formattedContent.replace(new RegExp(`{{${key}}}`, 'g'), value as string);
    }
    return formattedContent;
  }

  async stream(inputs: IPromptInput): Promise<IAsyncIterableStream<string>> {
    const formattedPrompt = this.format(inputs);
    const { textStream } = await streamText({
      model: openai(this.configuration.modelName),
      prompt: formattedPrompt,
      temperature: this.configuration.temperature,
      maxTokens: this.configuration.maxTokens,
      topP: this.configuration.topP,
      frequencyPenalty: this.configuration.frequencyPenalty,
      presencePenalty: this.configuration.presencePenalty
    });
    return textStream as IAsyncIterableStream<string>;
  }

  async execute(inputs: IPromptInput): Promise<IPromptOutput> {
    if (this.outputType === 'structured') {
      const formattedPrompt = this.format(inputs);
      const schema = z.object(this.outputSchema as z.ZodRawShape);
      const { object } = await generateObject({
        model: openai(this.configuration.modelName),
        schema,
        prompt: formattedPrompt,
        temperature: this.configuration.temperature,
        maxTokens: this.configuration.maxTokens,
        topP: this.configuration.topP,
        frequencyPenalty: this.configuration.frequencyPenalty,
        presencePenalty: this.configuration.presencePenalty
      });
      return object as IPromptOutput;
    } else {
      const { text } = await generateText({
        model: openai(this.configuration.modelName),
        prompt: this.format(inputs),
        temperature: this.configuration.temperature,
        maxTokens: this.configuration.maxTokens,
        topP: this.configuration.topP,
        frequencyPenalty: this.configuration.frequencyPenalty,
        presencePenalty: this.configuration.presencePenalty
      });
      return { text };
    }
  }

  updateMetadata(props: { metadata: Partial<IPromptModel['metadata']> }): void {
    this.metadata = { ...this.metadata, ...props.metadata };
    this.metadata.lastModified = new Date().toISOString();
  }

  getSummary(): string {
    return `${this.name} (${this.category}): ${this.description}`;
  }

  async save(): Promise<void> {
    await this.fileSystem.savePrompt({ promptData: this });
    this._isSaved = true;
  }

  get inputZodSchema(): z.ZodObject<IPromptInput> {
    return z.object(this.inputSchema as z.ZodRawShape);
  }

  get outputZodSchema(): z.ZodObject<IPromptOutput> {
    return z.object(this.outputSchema as z.ZodRawShape);
  }

  async load(props: { filePath: string }): Promise<void> {
    const promptData = await this.fileSystem.loadPrompt({ category: this.category, promptName: this.name });
    Object.assign(this, promptData);
    this._isSaved = true;
  }

  versions(): string[] {
    // Assuming PromptFileSystem doesn't have a getVersions method, we'll need to implement this differently
    // For now, we'll return an empty array as a placeholder
    return [];
  }

  switchVersion(props: { version: string }): void {
    // Implement version switching logic
    // This is a placeholder and should be implemented based on your versioning strategy
    console.log(`Switching to version ${props.version}`);
  }

  get isSaved(): boolean {
    return this._isSaved;
  }

  static async loadPromptByName(name: string, fileSystem: PromptFileSystem): Promise<PromptModel> {
    const [category, promptName] = name.split('/');
    const promptData = await fileSystem.loadPrompt({ category, promptName });
    const prompt = new PromptModel(promptData as IPromptModelRequired, fileSystem);
    prompt.isLoadedFromStorage = true;
    return prompt;
  }

  static async promptExists(name: string, fileSystem: PromptFileSystem): Promise<boolean> {
    const [category, promptName] = name.split('/');
    return fileSystem.promptExists({ category, promptName });
  }

  static async listPrompts(category?: string, fileSystem?: PromptFileSystem): Promise<string[]> {
    if (!fileSystem) {
      throw new Error("FileSystem is required for listing prompts");
    }
    return fileSystem.listPrompts({ category });
  }
}
