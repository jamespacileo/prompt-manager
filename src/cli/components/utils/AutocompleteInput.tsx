import type React from "react";
import { useState, useEffect } from "react";
import { Box, Text, useInput } from "ink";
import GridOptions, { type Option } from "./GridOptions";

interface AutocompleteInputProps {
	options: Option[];
	onSelect: (option: Option) => void;
	placeholder?: string;
	maxDisplayedOptions?: number;
	descriptionMaxLength?: number;
	isFocused?: boolean;
}

const AutocompleteInput: React.FC<AutocompleteInputProps> = ({
	options,
	onSelect,
	placeholder = "Type to search...",
	maxDisplayedOptions = 5,
	descriptionMaxLength = 50,
	isFocused = true,
}) => {
	const [inputValue, setInputValue] = useState("");
	const [filteredOptions, setFilteredOptions] = useState(options);
	const [selectedIndex, setSelectedIndex] = useState(0);

	useEffect(() => {
		const filtered = options.filter(
			(option) =>
				option.label.toLowerCase().includes(inputValue.toLowerCase()) ||
				option.value.toLowerCase().includes(inputValue.toLowerCase()),
		);
		setFilteredOptions(filtered);
		setSelectedIndex(0);
	}, [inputValue, options]);

	useInput(
		(input, key) => {
			if (key.return) {
				onSelect(filteredOptions[selectedIndex]);
			} else if (key.upArrow) {
				setSelectedIndex((prev) => (prev > 0 ? prev - 1 : prev));
			} else if (key.downArrow) {
				setSelectedIndex((prev) =>
					prev < filteredOptions.length - 1 ? prev + 1 : prev,
				);
			} else if (key.backspace) {
				setInputValue((prev) => prev.slice(0, -1));
			} else if (!key.ctrl && !key.meta) {
				setInputValue((prev) => prev + input);
			}
		},
		{
			isActive: isFocused,
		},
	);

	return (
		<Box flexDirection="column">
			<Box>
				<Text>‚ùØ </Text>
				<Text>{inputValue || placeholder}</Text>
			</Box>
			<GridOptions
				options={filteredOptions}
				selectedIndex={selectedIndex}
				maxDisplayedOptions={maxDisplayedOptions}
				descriptionMaxLength={descriptionMaxLength}
			/>
		</Box>
	);
};

export default AutocompleteInput;
