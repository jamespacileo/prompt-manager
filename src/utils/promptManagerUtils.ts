import path from 'path';
import { IPrompt, IPromptInput, IPromptOutput } from '../types/interfaces';
import { PromptModel } from '../promptModel';

export function getPromptFilePath(basePath: string, category: string, name: string): string {
  return path.join(basePath, category, name, 'prompt.json');
}

export function validateCategoryAndName(category: string, name: string): void {
  if (!category || !name || category.trim() === '' || name.trim() === '') {
    throw new Error('Prompt category and name are required and cannot be empty');
  }
}

export function mapPromptToFileInfo(prompt: PromptModel<any, any>, basePath: string): IPrompt<IPromptInput, IPromptOutput> & { filePath: string } {
  const promptData = prompt as unknown as IPrompt<IPromptInput, IPromptOutput>;
  return {
    ...promptData,
    filePath: promptData.category && promptData.name
      ? getPromptFilePath(basePath, promptData.category, promptData.name)
      : ''
  };
}

export function handlePromptNotFound(category: string, name: string): never {
  throw new Error(`Prompt "${name}" in category "${category}" does not exist`);
}
