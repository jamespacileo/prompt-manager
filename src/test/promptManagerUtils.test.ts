import path from 'path';
import { describe, it, expect } from 'bun:test';
import { getPromptFilePath, validateCategoryAndName, mapPromptToFileInfo, handlePromptNotFound } from '../utils/promptManagerUtils';
import { PromptModel } from '../promptModel';

describe('promptManagerUtils', () => {
  describe('getPromptFilePath', () => {
    it('should return the correct file path', () => {
      const result = getPromptFilePath('/base/path', 'category', 'name');
      expect(result).toEqual('/base/path/category/name/prompt.json');
    });
  });

  describe('validateCategoryAndName', () => {
    it('should not throw for valid inputs', () => {
      expect(() => validateCategoryAndName('category', 'name')).not.toThrow();
    });

    it('should throw for empty category', () => {
      expect(() => validateCategoryAndName('', 'name')).toThrow('Prompt category and name are required and cannot be empty');
    });

    it('should throw for empty name', () => {
      expect(() => validateCategoryAndName('category', '')).toThrow('Prompt category and name are required and cannot be empty');
    });
  });

  describe('mapPromptToFileInfo', () => {
    it('should map prompt to file info', () => {
      const mockPrompt = {
        category: 'testCategory',
        name: 'testName',
        description: 'Test description',
        template: 'Test template',
        version: '1.0.0',
      } as PromptModel<any, any>;

      const result = mapPromptToFileInfo(mockPrompt, '/base/path');
      expect(result).toMatchSnapshot();
    });
  });

  describe('handlePromptNotFound', () => {
    it('should throw an error with the correct message', () => {
      expect(() => handlePromptNotFound('category', 'name')).toThrow('Prompt "name" in category "category" does not exist');
    });
  });
});
