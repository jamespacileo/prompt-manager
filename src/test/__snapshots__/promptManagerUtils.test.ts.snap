// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`promptManagerUtils mapPromptToFileInfo should map prompt to file info 1`] = `
{
  "category": "testCategory",
  "description": "Test description",
  "filePath": "/base/path/testCategory/testName/prompt.json",
  "name": "testName",
  "template": "Test template",
  "version": "1.0.0",
}
`;
